payments/providers/bambora_payform.py

```diff
@@ -254,6 +264,102 @@ def payload_add_auth_code(self, payload: dict):
         data = "{}|{}".format(payload["api_key"], payload["order_number"])
         payload.update(authcode=self.calculate_auth_code(data))
 
+    def initiate_refund(self, order: Order) -> OrderRefund:
+        # Orders which are PAID_MANUALLY may not have an entry in Bambora,
+        # so we can't guarantee that the refund will be executed
+        if order.status != OrderStatus.PAID:
+            raise ValidationError(
+                "Cannot refund an order that has not been paid through VismaPay"
+            )
+
+        order_token = (
+            order.tokens.exclude(token__isnull=True, token__iexact="")
+            .order_by("-created_at")
+            .first()
+        )
+
+        if not order_token:
+            raise ValidationError(
+                "Cannot refund an order that has not been paid through VismaPay"
+            )
+
+        payload = {
+            "version": "w3.1",
+            "api_key": self.config.get(VENE_PAYMENTS_BAMBORA_API_KEY),
+            "amount": price_as_fractional_int(order.price),
+            "notify_url": self.get_notify_refund_url(),
+            "order_number": f"{order.order_number}-{order_token.created_at.timestamp()}",
+        }
+        self.payload_add_auth_code(payload)
+
+        try:
+            r = requests.post(self.url_refund, json=payload, timeout=60)
+            r.raise_for_status()
+            response = r.json()
+            result = response["result"]
+
+            if result == 1:
+                raise PayloadValidationError(
+                    f"{_('Payment payload data validation failed: ')} {' '.join(response['errors'])}"
+                )
+            elif result == 10:
+                raise ServiceUnavailableError(
+                    _("Payment service is down for maintenance")
+                )
+
+            return OrderRefund.objects.create(
+                order=order, amount=order.price, refund_id=str(response["refund_id"])
+            )
+        except RequestException as e:
+            raise ServiceUnavailableError(_("Payment service is unreachable")) from e
+
+    def handle_notify_refund_request(self) -> HttpResponse:
+        request = self.request
+        logger.debug(
+            "Handling Bambora notify refund request, params: {}.".format(request.GET)
+        )
+
+        refund_id = request.GET.get("REFUND_ID")
+
+        try:
+            refund = OrderRefund.objects.get(refund_id=refund_id)
+            order = refund.order
+        except OrderRefund.DoesNotExist:
+            # Target order might be deleted after posting but before the notify arrives
```

Comment:
If someone creates a refund manually through Bambora UI (without making a OrderRefund at vene db first), does Bambora make a http request to notify us? Are the admins going to use Bambora UI to make refunds manually, and is that something that needs to be supported?

Reply:
> _does Bambora make a http request to notify us?_

No

> _Are the admins going to use Bambora UI to make refunds manually, and is that something that needs to be supported?_

As I understand, they are not going to use Bambora to make the refunds, I don't think they're doing it right now either. So they would only use the Admin UI to handle the refunds.